ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -373,6 +391,57 @@ void fillChunks() {
   }
 
   /**
+   * Submits encrypt job for the given {@link PutChunk} and processes the callback for the same
+   * @param putChunk the {@link PutChunk} that needs to be encrypted
+   * @param encryptContent {@code true} if content needs to be encrypted. {@code false} otherwise
+   */
+  private void encryptChunk(PutChunk putChunk, boolean encryptContent) {
+    try {
+      logger.trace("Chunk at index {} state moves to {}", putChunk.chunkIndex, ChunkState.Encrypting);
+      putChunk.state = ChunkState.Encrypting;
+      if (encryptContent) {
+        putChunk.buf.flip();
+        putChunk.chunkBlobSize = putChunk.buf.remaining();
+      }
+      logger.trace("Submitting encrypt job for chunk at index " + putChunk.chunkIndex);
+      exec.submitJob(new EncryptJob(passedInBlobProperties.getAccountId(), passedInBlobProperties.getContainerId(),
+          encryptContent ? putChunk.buf : null, ByteBuffer.wrap(putChunk.chunkUserMetadata), kms.getRandomKey(),
+          cryptoService, kms, (EncryptJob.EncryptJobResult result, Exception exception) -> {
+        logger.trace("Processing encrypt job callback for chunk at index " + putChunk.chunkIndex);
+        if (exception == null && !isOperationComplete()) {
+          if (encryptContent) {
+            putChunk.buf = result.getEncryptedBlobContent();
+          }
+          putChunk.encryptedPerBlobKey = result.getEncryptedKey();
+          putChunk.chunkUserMetadata = result.getEncryptedUserMetadata().array();
+          logger.trace("Completing encrypt job result for chunk at index " + putChunk.chunkIndex);
+          putChunk.onFillComplete(true);
```

Comment:
this is more like `onEncryptionComplete()` - `onFillComplete()` seems a little off here.

Reply:
Yes. But onFillingComplete() is expected to be called when the chunk is fully filled and ready to move to next state. W/ encryption inplay, we have to call the same method just that we wait until encryption is complete before we call it. Let me know your thoughts on how to improve this. One option is to introduce a method called onEncryptionComplete(), but the impl will just be onFillComplete() or pretty much do what onFilleComplete() does. will sync up with you offline.